text/3711-required-inlining.md

```diff
@@ -0,0 +1,153 @@
+- Feature Name: `inline(must)` and `inline(required)`
+- Start Date: 2024-09-17
+- RFC PR: [rust-lang/rfcs#3711](https://github.com/rust-lang/rfcs/pull/3711)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+rustc supports inlining functions with the `#[inline]`/`#[inline(always)]`/
+`#[inline(never)]` attributes, but these are only hints to the compiler.
+Sometimes it is desirable to require inlining, not just suggest inlining:
+security APIs can depend on being inlined for their security properties, and
+performance intrinsics can work poorly if not inlined.
+
+# Motivation
+[motivation]: #motivation
+
+rustc's current inlining attributes are only hints, they do not guarantee that
+inlining is performed and users need to check manually whether inlining was
+performed. While this is desirable for the vast majority of use cases, there are
+circumstances where inlining is necessary to maintain the security properties of
+an API, or where performance of intrinsics is likely to suffer if not inlined.
+For example, consider:
+
+- Armv8.3-A's pointer authentication, when used explicitly with intrinsics[^1],
+rely on being inlined to guarantee their security properties. Rust should make
+it easier to guarantee that these intrinsics will be inlined and emit an error
+if that is not possible or has not occurred.
+- Arm's Neon and SVE performance intrinsics have work poorly if not inlined.
+Since users rely on these intrinsics for their application's performance, Rust
+should be able to warn users when these have not been inlined and performance
+will not be as expected.
+
+In general, for many intrinsics, inlining is not an optimisation but an
+important part of the semantics, regardless of optimisation level.
```

Comment:
If you truly do mean that inlining is part of the semantics, then I think a lot of work needs to be done here, including making this "inlining" happen for the MIR that Miri executes.

Reply:
Assuming I'm reading this and the draft implementation right, `-Zmir-enable-passes=-Inline` in the draft PR is unsound.

Reply:
I think my wording was too strong here, there wouldn't necessarily be unsoundness, but for something like SIMD intrinsics, they basically only make sense if you inline them - it isn't wrong if you don't, but it isn't more debuggable or anything like that, just slower. These intrinsics only make sense inlined.